### PR TITLE
mpc code gen

### PR DESCRIPTION
**Pull request by ianwilkietomasik**
Notes:
- Y2A and Y2B giving bad values, so I removed these tests
- Placing array tests in a separate PR due to protocol selection and runtime issues testing mpc arrays

### CODE REVIEW

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/aby/ABYCodeGenerator.kt [hunk 1]
```diff
@@ -0,0 +1,796 @@
...
+
+    override fun output(protocol: Protocol, stmt: OutputNode): CodeBlock =
+        throw UnsupportedOperationException("cannot perform I/O in non-local protocol: ${stmt.toDocument().print()}")
+
+    override fun guard(protocol: Protocol, expr: AtomicExpressionNode): CodeBlock =
+        throw UnsupportedOperationException("ABY: Cannot execute conditional guard: ${expr.toDocument().print()}")
+
+    private fun roleToCodeBlock(role: Role): CodeBlock =
+        when (role) {
+            Role.CLIENT -> CodeBlock.of("%T.CLIENT", Role::class.asClassName())
```
Comment (cacay) [inline compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/aby/ABYCodeGenerator.kt:651]: No need to case here. Copy what you do in `protocolToAbyPartyCircuit`.

**Review by cacay — CHANGES_REQUESTED** [review]
Mostly minor style suggestions. You should merge in master because there are some breaking changes that prevent auto merge.

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/aby/ABYCodeGenerator.kt [hunk 2]
```diff
@@ -0,0 +1,796 @@
...
+                        conversion
+                    )
+                } else {
+                    CodeBlock.of("%N", context.kotlinName(expr.temporary.value, protocol))
+                }
+            }
+
+            is OperatorApplicationNode -> {
+                when (expr.operator) {
+                    is BinaryOperator ->
```
[the remark this answers is not in the document]
Comment (ianwilkietomasik) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/aby/ABYCodeGenerator.kt:457]: I'm not exactly seeing the idea you have for this, would be interested in a bit more clarification, thanks!

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/CodeGenerationBackend.kt [hunk 3]
```diff
@@ -1,7 +1,24 @@
...
 import edu.cornell.cs.apl.viaduct.backends.commitment.CommitmentBackend
+import edu.cornell.cs.apl.viaduct.selection.ProtocolFactory
+import edu.cornell.cs.apl.viaduct.selection.unions
+import edu.cornell.cs.apl.viaduct.syntax.intermediate.ProgramNode
 
 /** Combines all back ends that support code generation. */
-val CodeGenerationBackend: Backend = listOf(CleartextBackend, CommitmentBackend).unions()
+object CodeGenerationBackend : Backend by backends.unions() {
+    override fun protocolFactory(program: ProgramNode): ProtocolFactory {
+        // TODO: fix ABYProtocolFactory to get rid of this hack
```
Comment (rolph-recto) [inline compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/CodeGenerationBackend.kt:14]: what is the fix for `ABYProtocolFactory`?

File: runtime/src/main/kotlin/edu/cornell/cs/apl/viaduct/runtime/aby/Utility.kt [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+        currentShare = mux
+    }
+    return currentShare
+}
+
+fun Array<Share>.secretUpdateModify(circuit: Circuit, index: Share, operation: (Share) -> Share) {
+    for (i in this.indices) {
+        val rhs = operation(this[i])
+        val guard = circuit.putEQGate(index, circuit.putCONSGate(i.toBigInteger(), 32))
+        val mux = circuit.putMUXGate(this[i], rhs, guard) // TODO() - check arg order
```
Comment (rolph-recto) [inline runtime/src/main/kotlin/edu/cornell/cs/apl/viaduct/runtime/aby/Utility.kt:32]: remove these TODOs

File: runtime/src/main/kotlin/edu/cornell/cs/apl/viaduct/runtime/aby/Utility.kt [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+        this[i] = mux
+    }
+}
+
+fun Array<Share>.secretUpdateSet(circuit: Circuit, index: Share, argument: Share) {
+    for (i in this.indices) {
+        val guard = circuit.putEQGate(index, circuit.putCONSGate(i.toBigInteger(), 32))
+        val mux = circuit.putMUXGate(this[i], argument, guard) // TODO() - check arg order
```
Comment (rolph-recto) [inline runtime/src/main/kotlin/edu/cornell/cs/apl/viaduct/runtime/aby/Utility.kt:40]: remove TODO

File: compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/CodeGenerationBackend.kt (hunk 3 above)
Comment (cacay) [reply compiler/src/main/kotlin/edu/cornell/cs/apl/viaduct/backends/CodeGenerationBackend.kt:14]: That TODO is copied from code that I wrote. `ABYProtocolFactory` should not need a reference to the parent factory. I think we discussed a change that would make that possible, but I don't quite remember what it was.

**Comment by cacay** [comment]
This also fixes #346.

**Pull request merged**